examples/common/convert_c4.py

```diff
@@ -4,105 +4,344 @@
+                                                   name='en',
+                                                   split=split,
+                                                   streaming=True)
 
-    return args.parse_args()
+    def __iter__(self) -> Iterable[Dict[str, bytes]]:
+        for sample in self.hf_dataset:
+            # convert to bytes to store in MDS binary format
+            yield {'text': sample['text'].encode('utf-8')}
 
 
-def build_hf_c4_dataset(split: str) -> IterableDataset:
-    """Collect the samples for this dataset split.
+class ConcatTextC4(IterableDataset):
+    """An IterableDataset that returns text samples for MDSWriter.
+
+    Returns dicts of {'text': bytes}
+    """
+
+    def __init__(self, split: str, max_length: int, bos_text: str,
+                 eos_text: str, no_wrap: bool):
+        self.max_length = max_length
+        self.bos_text = bos_text
+        self.eos_text = eos_text
+        self.should_wrap = not no_wrap
+        self.hf_dataset = hf_datasets.load_dataset(path='c4',
+                                                   name='en',
+                                                   split=split,
+                                                   streaming=True)
+
+    def __iter__(self) -> Iterable[Dict[str, bytes]]:
+        buffer = ''
+        for sample in self.hf_dataset:
+            buffer = buffer + self.bos_text + sample['text'] + self.eos_text
+            while len(buffer) >= self.max_length:
+                concat_sample = buffer[:self.max_length]
+                buffer = buffer[self.max_length:] if self.should_wrap else ''
+                # convert to bytes to store in MDS binary format
+                yield {'text': concat_sample.encode('utf-8')}
+
+
+class ConcatTokensC4(IterableDataset):
+    """An IterableDataset that returns token samples for MDSWriter.
+
+    Returns dicts of {'tokens': bytes}
+
+    To use data created by this class and written to MDS format:
+
+    ```python
+        import torch
+        from streaming.base import StreamingDataset
+        from transformers import AutoTokenizer
+
+        tokenizer = AutoTokenizer.from_pretrained('your/tokenizer')
+        ds = StreamingDataset(local='mds-data-folder', split='val')
+
+        # note, you need to copy the numpy array because the original is non-writeable
+        # and torch does not support non-writeable tensors, so you get a scary warning and
+        # if you do try to write to the tensor you get undefined behavior
+        tokens = torch.from_numpy(np.frombuffer(ds[0]['tokens'], dtype=np.int64).copy())
+        print(tokenizer.decode(tokens))
+    ```
+    """
+
+    def __init__(
+        self,
+        split: str,
+        tokenizer: PreTrainedTokenizerBase,
+        max_length: int,
+        bos_text: str,
+        eos_text: str,
+        no_wrap: bool,
+    ):
+        self.tokenizer = tokenizer
+        os.environ['TOKENIZERS_PARALLELISM'] = 'false'
```

Comment:
Should this be set in `NoConcatC4` and `ConcatTextC4` as well?

Reply:
I think there are no tokenizers there, so not needed